tools/database/src/resharding.rs

```diff
@@ -0,0 +1,188 @@
+use near_store::db::{MixedDB, ReadOrder, RecoveryDB, RocksDB, SplitDB};
+use near_store::genesis::initialize_sharded_genesis_state;
+use near_store::{Mode, NodeStorage, Store, Temperature};
+use nearcore::NightshadeRuntimeExt;
+use nearcore::{open_storage, NearConfig, NightshadeRuntime};
+
+#[derive(clap::Args)]
+pub(crate) struct ReshardingCommand {
+    /// The block height at which resharding is performed.
+    /// This should be, usually, the block before shard layout has changed.
+    #[clap(long)]
+    height: BlockHeight,
+
+    /// The shard id before resharding.
+    #[clap(long)]
+    shard_id: ShardId,
+
+    /// Path to write the new trie nodes created by the resharding operation.
+    #[clap(long, group("output"))]
+    write_path: Option<PathBuf>,
+
+    /// Restore potentially missing trie nodes in cold database. This operation is idempotent.
+    #[clap(long, group("output"))]
+    restore: bool,
+}
+
+impl ReshardingCommand {
+    pub(crate) fn run(&self, mut config: NearConfig, home_dir: &Path) -> anyhow::Result<()> {
+        Self::check_resharding_config(&mut config);
+
+        let mut chain = self.get_chain(config, home_dir)?;
+
+        let block_hash = *chain.get_block_by_height(self.height)?.hash();
+
+        let resharding_request = chain.custom_build_state_for_resharding_preprocessing(
+            &block_hash,
+            &block_hash,
+            self.shard_id,
+        )?;
+
+        let shard_uid = resharding_request.shard_uid;
+
+        let response = Chain::build_state_for_split_shards(resharding_request);
+        let ReshardingResponse { sync_hash, new_state_roots: state_roots, .. } = response;
+
+        let state_roots = state_roots?;
+        tracing::info!(target: "resharding", ?state_roots, "state roots");
+
+        chain.custom_build_state_for_split_shards_postprocessing(
+            shard_uid,
+            &sync_hash,
+            state_roots,
+        )?;
+
+        if self.restore {
+            // In restore mode print database write statistics.
+            chain.runtime_adapter.store().get_store_statistics().map(|stats| {
+                stats.data.iter().for_each(|(metric, values)| {
+                    tracing::info!(target: "resharding", %metric, ?values);
+                })
+            });
+        }
+
+        Ok(())
+    }
+
+    fn get_store(&self, home_dir: &Path, config: &mut NearConfig) -> Result<Store, anyhow::Error> {
+        // Open hot and cold as usual.
+        let storage = open_storage(home_dir, config)?;
+
+        let cold_db = storage
+            .cold_db()
+            .expect("resharding tool can be used only on archival nodes")
+            .to_owned();
+        let hot_db = storage.into_inner(Temperature::Hot);
+
+        // We need real split db so that it correctly handles reads of missing
+        // values in the columns that are not in the cold db.
+        let split_db = SplitDB::new(hot_db, cold_db.clone());
```

Comment:
I wonder if we can use cold db exclusively in order to save some time by not having to read from hot first. I'm not sure if that would actually help but it's something to consider given it's a long operation and any speedup would help. Just a suggestion for future improvement in case it's needed.

Reply:
I can try using mixeddb directly with priority on cold db, maybe it will speed up a bit the operation